cpp/arcticdb/vector_db/vector_db.cpp

```diff
@@ -0,0 +1,102 @@
+        // top k, not the top k+1.
+        auto lp = 2;
+        // Let's pretend that only the lp-norms exist. We'll use p=2 for now as a default.
+
+        procs.broadcast(
+                [&store, &top_k, lp, this](const ProcessingSegment &proc) {
+                    for (const auto& slice_and_key: proc.data()) {
+                        std::vector <std::shared_ptr<Column>> columns = slice_and_key.segment(store).columns();
+                        for (auto&& [idx, col]: folly::enumerate(columns)) {
+                            internal::check<ErrorCode::E_ASSERTION_FAILURE>(
+                                    static_cast<long unsigned>(col->row_count()) == this->query_vector_.size(),
+                                    "Expected vector of length {}, got vector of length {}.",
+                                    col->row_count(),
+                                    this->query_vector_.size());
+                            col->type().visit_tag([&top_k, &slice_and_key, &store, &col=col, idx=idx, lp, this](auto type_desc_tag) {
+                                using TypeDescriptorTag = decltype(type_desc_tag);
+                                using RawType = typename TypeDescriptorTag::DataTypeTag::raw_type;
+
+                                if constexpr (is_floating_point_type(TypeDescriptorTag::DataTypeTag::data_type)) {
+
+                                    ColumnData col_data = col->data();
+                                    double sum_differences_exponentiated = 0;
+                                    auto j = 0u;
+
+                                    while (auto block = col_data.next<TypeDescriptorTag>()) {
+                                        auto ptr = reinterpret_cast<const RawType *>(block.value().data());
+                                        for (auto progress_in_block = 0u; progress_in_block <
+                                                                          block.value().row_count(); ++progress_in_block, ++j, ++ptr) {
+                                            sum_differences_exponentiated += pow(
+                                                    abs(*ptr - this->query_vector_[j]), lp);
+                                        }
+                                    }
+
+                                    top_k.try_insert(NamedColumnSortedByDistance(
+                                            sum_differences_exponentiated,
+                                            col,
+                                            slice_and_key.segment(store).field(idx).name()));
+                                }
+                            });
+                        }
+                    }
+                }
+        );
+        // By this time we should have iterated through each block of each SegmentInMemory of each ProcessingSegment;
+        // and in each case read a vector and inserted it if more similar than the running top_k_. Now we write to a new
```

Comment:
After review I'll rephrase this comment (which surely can't cause CI problems...) to something like 'By this time we've read each vector, keeping a running count of the top-k vectors closest to the query vector of the vectors read so far. Now we write to a new segment.'

Reply:
Going to push some changes given William's comments below so I'll do this too.